stumpy/scrump.py

```diff
@@ -103,67 +109,125 @@ def _compute_PI(
-            P_squared[thread_idx][mask] = squared_distance_profile[mask]
-            I[thread_idx][mask] = i
-
-        I[thread_idx, i] = np.argmin(squared_distance_profile)
-        P_squared[thread_idx, i] = squared_distance_profile[I[thread_idx, i]]
-        if P_squared[thread_idx, i] == np.inf:  # pragma: no cover
-            I[thread_idx, i] = -1
-        else:
-            j = I[thread_idx, i]
-            # Given the squared distance, work backwards and compute QT
-            QT_j = (m - P_squared[thread_idx, i] / 2.0) * (Σ_T[j] * σ_Q[i]) + (
-                m * M_T[j] * μ_Q[i]
+            core._apply_exclusion_zone(squared_distance_profile, i, excl_zone, np.inf)
+
+        nn_idx = np.argmin(squared_distance_profile)
+        if nn_idx not in I[thread_idx, i]:
+            # It is more than likely that the top-k values for the `i`-th subsequence
+            # will be already populated. So, we must shift-insert here
+            core._shift_insert_at_index(
+                P_squared[thread_idx, i], 0, squared_distance_profile[nn_idx]
+            )
+            core._shift_insert_at_index(I[thread_idx, i], 0, nn_idx)
+
+        if P_squared[thread_idx, i, 0] == np.inf:  # pragma: no cover
+            I[thread_idx, i, 0] = -1
+            continue
+
+        j = nn_idx
+        # Given the squared distance, work backwards and compute QT
+        QT_j = (m - P_squared[thread_idx, i, 0] / 2.0) * (Σ_T[j] * σ_Q[i]) + (
+            m * M_T[j] * μ_Q[i]
+        )
+        QT_j_prime = QT_j
+        # Update top-k for both subsequences `S[i+g] = T[i+g:i+g+m]`` and
+        # `S[j+g] = T[j+g:j+g+m]` (i.e., the right neighbors of `T[i : i+m]` and
+        # `T[j:j+m]`) by using the distance between `S[i+g]` and `S[j+g]`
+        for g in range(1, min(s, l - i, w - j)):
+            QT_j = (
+                QT_j
+                - T_B[j + g - 1] * T_A[i + g - 1]
+                + T_B[j + g + m - 1] * T_A[i + g + m - 1]
             )
-            QT_j_prime = QT_j
-            for k in range(1, min(s, l - i, w - j)):
-                QT_j = (
-                    QT_j
-                    - T_B[j + k - 1] * T_A[i + k - 1]
-                    + T_B[j + k + m - 1] * T_A[i + k + m - 1]
+            D_squared = core._calculate_squared_distance(
+                m,
+                QT_j,
+                M_T[j + g],
+                Σ_T[j + g],
+                μ_Q[i + g],
+                σ_Q[i + g],
+            )
+            if D_squared < P_squared[thread_idx, i + g, -1]:
+                idx = np.searchsorted(
+                    P_squared[thread_idx, i + g], D_squared, side="right"
+                )
+                # Due to numerical error, it is possible that the element that is
+                # about to insert at idx is identical to an element of array located
+                # at idx, idx + 1, .... Hence, we should traverse full array.
+                # This is optimized in the if conditon.
```

Comment:
Maybe remove this comment, or make it more clear by providing a simple example (?)

Reply:
I propose that we remove this comment and ensure that we have a unit test that will fail if anybody has the "bright idea" in the future to only check a portion of `I[thread_idx, i + g]`. Then, in that unit test, clearly explain what the test is for and why it is there.

To me, a properly constructed failing unit test is the best defense to prevent others from making the same bad assumption regarding precision. We have to make sure to add the same test to `scraamp` as well.

Reply:
Cool. I will work on this. Although I do not have the full picture of such test function in my mind as of now, I will will do my best to write it, and then we can revise it.